transcription/ibm.go

```diff
@@ -0,0 +1,134 @@
+package transcription
+
+import (
+	"bufio"
+	"encoding/base64"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+// IBMResult is the result of an IBM transcription. See
+// https://www.ibm.com/smarterplanet/us/en/ibmwatson/developercloud/doc/speech-to-text/output.shtml
+// for details.
+type IBMResult struct {
+	ResultIndex int              `json:"result_index"`
+	Results     []ibmResultField `json:"results"`
+}
+type ibmResultField struct {
+	Alternatives []ibmAlternativesField `json:"alternatives"`
+	Final        bool                   `json:"final"`
+}
+type ibmAlternativesField struct {
+	WordConfidence    []ibmWordConfidence `json:"word_confidence"`
+	OverallConfidence float64             `json:"confidence"`
+	Transcript        string              `json:"transcript"`
+	Timestamps        []ibmWordTimestamp  `json:"timestamps"`
+}
+type ibmWordConfidence []interface{}
+type ibmWordTimestamp []interface{}
+
+// TranscribeWithIBM transcribes a given audio file using the IBM Watson
+// Speech To Text API
+func TranscribeWithIBM(filePath string, IBMUsername string, IBMPassword string) (*IBMResult, error) {
+	result := new(IBMResult)
+
+	url := "wss://stream.watsonplatform.net/speech-to-text/api/v1/recognize?model=en-US_BroadbandModel"
+	header := http.Header{}
+	header.Set("Authorization", "Basic "+basicAuth(IBMUsername, IBMPassword))
+
+	dialer := websocket.DefaultDialer
+	ws, _, err := dialer.Dial(url, header)
+	if err != nil {
+		return nil, err
+	}
+	defer ws.Close()
+
+	requestArgs := map[string]interface{}{
+		"action":             "start",
+		"content-type":       "audio/flac",
+		"continuous":         true,
+		"word_confidence":    true,
+		"timestamps":         true,
+		"profanity_filter":   false,
+		"interim_results":    false,
+		"inactivity_timeout": -1,
```

Comment:
Basically, if there is x seconds of no speaking, we don't want IBM to end the session. If you look at IBM's own JavaScript demo, they also set the timeout to infinity.